umbral/openssl.py

```diff
@@ -0,0 +1,141 @@
+from cryptography.hazmat.backends.openssl import backend
+
+
+def _get_new_BN():
+    """
+    Returns a new and initialized OpenSSL BIGNUM.
+
+    TODO: Add flag for consttime
+    """
+    new_bn = backend._lib.BN_new()
+    backend.openssl_assert(new_bn != backend._ffi.NULL)
+    new_bn = backend._ffi.gc(new_bn, backend._lib.BN_clear_free)
+
+    return new_bn
+
+
+def _get_ec_group_by_curve_nid(curve_nid: int):
+    """
+    Returns the group of a given curve via its OpenSSL nid.
+    """
+    group = backend._lib.EC_GROUP_new_by_curve_name(curve_nid)
+    backend.openssl_assert(group != backend._ffi.NULL)
+
+    return group
+
+
+def _get_ec_order_by_curve_nid(curve_nid: int):
+    """
+    Returns the order of a given curve via its OpenSSL nid.
+    """
+    ec_group = _get_ec_group_by_curve_nid(curve_nid)
+    ec_order = _get_new_BN()
+
+    with backend._tmp_bn_ctx() as bn_ctx:
+        res = backend._lib.EC_GROUP_get_order(ec_group, ec_order, bn_ctx)
+        backend.openssl_assert(res == 1)
+
+    return ec_order
+
+
+def _get_ec_generator_by_curve_nid(curve_nid: int):
+    """
+    Returns the generator point of a given curve via its OpenSSL nid.
+    """
+    ec_group = _get_ec_group_by_curve_nid(curve_nid)
+
+    generator = backend._lib.EC_GROUP_get0_generator(ec_group)
+    backend.openssl_assert(generator != backend._ffi.NULL)
+
+    return generator
+
+
+def _bn_is_on_curve(check_bn, curve_nid: int):
+    """
+    Checks if a given OpenSSL BIGNUM is within the provided curve's order.
+    Returns True if the provided BN is on the curve, or False if the BN is zero
+    or not on the curve.
+    """
+    ec_order = _get_ec_order_by_curve_nid(curve_nid)
+
+    # is_zero workaround since BN_is_zero is not exposed yet.
+    is_zero = backend._lib.BN_cmp(check_bn, backend._int_to_bn(0))
```

Comment:
I think below this line I check that `is_zero == 1`, right?

Reply:
What I mean is that the name `is_zero` doesn't make sense. You are not testing if `check_bn == 0`, but getting a comparison that returns -1, 0 or 1.

Reply:
🤔 I haven't thought about it like that. Given the comment above it, it should be enough to determine, but I think I might prefer `bn_is_zero` over `greater_than_zero`. Thoughts?

Reply:
Ah OK, I see what you mean here. You intention is to validate that `check_bn` is different than 0, but you end up comparing if `check_bn > 0`. Both are equivalent for BIGNUMs. In any case, the name `is_zero` doesn't seem very appropriate, because you end up testing if `check_bn` is greater than 0.

Reply:
Specifically, I am computing whether this statement is true:
`0 < check_bn < ec_order`

How do you mean that `check_bn > 0` and `is_zero` are both equivalent?